### PR TITLE
Allowing Admin authenticated requests to bypass WebHook auth pipeline

### PR DESCRIPTION
**Pull request by mathewc**
This facilitates debugging of WebHook methods, since your requests don't have to satisfy all the stringent WebHook requirements - you can just call into your code. Also adding some e2e tests that go through our WebHost for e2e request processing. These tests exercise some of our samples.

### CODE REVIEW

File: sample/HttpTrigger-CSharp/run.csx [hunk 1]
```diff
@@ -1,6 +1,4 @@
-﻿#r "System.Web.Http"
```
Comment (mathewc) [inline sample/HttpTrigger-CSharp/run.csx]: Fabio said to delete this

File: src/WebJobs.Script.WebHost/App_Start/AutofacBootstrap.cs [hunk 2]
```diff
@@ -15,32 +13,12 @@ namespace WebJobs.Script.WebHost.App_Start
 {
     public static class AutofacBootstrap
     {
-        internal static void Initialize(ContainerBuilder builder)
+        internal static void Initialize(ContainerBuilder builder, WebHostSettings settings)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/AutofacBootstrap.cs]: This change allows the new SelfHost e2e tests to pass in config info

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 3]
```diff
@@ -40,29 +41,41 @@ public FunctionsController(WebScriptHostManager scriptHostManager, WebHookReceiv
...
-            }
-
             HttpResponseMessage response = null;
             if (isWebHook)
             {
-                // This is a WebHook request so define a delegate for the user function.
-                // The WebHook Receiver pipeline will first validate the request fully
-                // then invoke this callback.
-                Func<HttpRequestMessage, Task<HttpResponseMessage>> invokeFunction = async (req) =>
+                if (authorizationLevel == AuthorizationLevel.Admin)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: This is the main change - here for Admin requests we dispatch directly to the user method

File: test/WebJobs.Script.Tests/SamplesEndToEndTests.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+using System.Web.Http;
+using WebJobs.Script.WebHost;
+using Xunit;
+
+namespace WebJobs.Script.Tests
+{
+    public class SamplesEndToEndTests : IClassFixture<SamplesEndToEndTests.TestFixture>
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/SamplesEndToEndTests.cs]: These tests use WebApi SelfHost to run tests against our WebHost

File: test/WebJobs.Script.Tests/SamplesEndToEndTests.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
+
+            HttpResponseMessage response = await this._fixture.Client.SendAsync(request);
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            string body = await response.Content.ReadAsStringAsync();
+            Assert.Equal("WebHook processed successfully! Foobar", body);
+        }
+
+        [Fact]
+        public async Task GenericWebHook_Post_AdminKey_Succeeds()
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/SamplesEndToEndTests.cs]: This is the test for the main change I'm making

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 6]
```diff
@@ -40,29 +41,41 @@ public FunctionsController(WebScriptHostManager scriptHostManager, WebHookReceiv
                 return new HttpResponseMessage(HttpStatusCode.NotFound);
             }
 
-            // Authorize the request
+            // Determine the authorization level of the request
             SecretManager secretManager = (SecretManager)controllerContext.Configuration.DependencyResolver.GetService(typeof(SecretManager));
+            AuthorizationLevel authorizationLevel = AuthorizationLevelAttribute.GetAuthorizationLevel(request, secretManager, functionName: function.Name);
```
Comment (davidebbo) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: `requestAuthorizationLevel`

**Comment by davidebbo** [comment]
:shipit:

**Comment by mathewc** [comment]
Merged

**Pull request closed without merging**